### DISCUSSION TITLE
What is the best approach to print from both cores?

### DISCUSSION THREAD

**Discussion by joba-1** [earlephilhower/arduino-pico#1277] [Q&A]
I noticed that output is not complete and in strange order if I do Serial.print()s from both cores, even if I‘m sure they don’t print at the same time.
Do I have to push messages in a queue on one core and pull/print on the other?

**Comment by earlephilhower** [comment]
A simple mutex controlled by your app, and making sure you `Serial.flush()` before releasing it, would suffice and not need any core in particular to do anything.  One core will then busy wait until the other is done and then take control.

FWIW, though, an individual `Serial.write()` should be protected already.  If you have multiple `Serial.puts()`/etc., though, then only a higher app-level mux like above could keep things unwound.  See the multicore example, the 2 cores never stomp on each other on a line-by-line basis because of this.
